snuba/web/query.py

```diff
@@ -234,9 +262,41 @@ def log_query_and_update_stats(
     return stats
 
 
-@split_query
 def parse_and_run_query(
     dataset: Dataset, request: Request, timer: Timer
+) -> ClickhouseQueryResult:
+    """
+    Runs a query, then records the results including metadata about each split.
+    """
+    query_list: List[ClickhouseQueryMetadata] = []
+    try:
+        result = _run_query(
+            dataset=dataset, request=request, timer=timer, query_list=query_list
+        )
+    except RawQueryException as error:
+        raise error
+    finally:
+        if settings.RECORD_QUERIES:
+            # send to redis
+            state.record_query(
+                {
+                    "request": request.body,
+                    "referrer": http_request.referrer,
+                    "timing": query_list[-1].timer,
+                    "status": query_list[-1].status,
```

Comment:
This code path now hinges on the assumption that `log_query_and_update_stats` is called at least once   during the execution of `_run_query`, and is susceptible to an `IndexError` if `query_list` isn't populated for some reason. This is a dangerous thing. It does appear that all of the normal return cases call `log_query_and_update_stats` at some point, but it's better to avoid that dependency if you can.

The fact that this is happening in a `finally` block makes this even trickier, since this is going to get called on _any_ exception — which could happen at any point during the function execution. At that point, this is effectively going to happen:

```
>>> try:
...   raise Exception('inner')
... finally:
...   raise Exception('outer')
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
Exception: inner

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
Exception: outer
```

This will also not propagate the outer exception, so if `RawQueryException` gets raised without updating stats, the request will 500 instead of returning whatever context was available in the other exception:

```
>>> try:
...   try:
...     raise Exception('inner')
...   finally:
...     raise Exception('outer') 
... except Exception as e:
...   print(e)
... 
outer
```

Reply:
Hmm, yeah. Updated this to not use finally and only attempt logging on RawQueryException / success